firewood/src/nibbles.rs

```diff
@@ -140,8 +140,23 @@ impl<'a, const LEADING_ZEROES: usize> DoubleEndedIterator for NibblesIterator<'a
     }
 }
 
+pub(crate) trait IntoBytes: Iterator<Item = u8> {
+    fn nibbles_into_bytes(&mut self) -> Vec<u8> {
+        let mut data = Vec::with_capacity(self.size_hint().0 / 2);
+
+        while let (Some(hi), Some(lo)) = (self.next(), self.next()) {
+            data.push((hi << 4) + lo);
+        }
+
+        data
+    }
+}
+impl<T: Iterator<Item = u8>> IntoBytes for T {}
```

Comment:
I think this would be great if we had `Item = Nibble` where `#[repr(transparent)] struct Nibble(u8)` (where you can only create the nibble if it's within the correct bounds).

Instead, I think something like this would make the call more explicit

```rust
/// CAUTION: only use with nibble iterators  
pub(crate) fn nibbles_into_bytes<Iter: Iterator<Item = u8>>(iter: Iter) -> Vec<u8> {
    let mut data = Vec::with_capacity(self.size_hint().0 / 2);

    while let (Some(hi), Some(lo)) = (self.next(), self.next()) {
        data.push((hi << 4) + lo);
    }

    data
}
```

Reply:
Created #440 for this larger change